### ISSUE TITLE
USE_REDIS_CACHE with Django 4.0's native Redis backend?

### ISSUE THREAD

**Issue by rstmsn** [rq/django-rq#571]
Django has [native support](https://docs.djangoproject.com/en/4.1/topics/cache/#redis) for Redis since version 4.0 via the ```django.core.cache.backends.redis.RedisCache``` backend.

If I create a Django cache, as follows:

```
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.redis.RedisCache',
        'LOCATION': 'redis://127.0.0.1:6379',
    }
}
```

Is it possible to configure ```django-rq``` to use this redis connection, using the ```USE_REDIS_CACHE``` option?

**Comment by selwin** [comment]
@rstmsn @obi-jerome I'd love to have a PR to implement this.

**Comment by mohamed99elsokary** [comment]
@selwin @rstmsn @obi-jerome  can you guys please give me a hint "What is that changes?"

**Comment by jerome-tgtg** [comment]
@mohamed99elsokary 
Before Django 4.0 we used django-redis to get a redis cache.
And django-rq could be simply configured then by setting "USE_REDIS_CACHE": "default".
Now we configure our cache with the new native Django Redis cache backend `django.core.cache.backends.redis.RedisCache`
But it is not declared as supported by django-rq, so we configured Redis manually for django-rq.

The native cache backend allows us to have one less dependency and to rely on the work of the Django Project that will probably give us a cache backend that is more reliable, integrated and performant, I guess.

It would be cool if django-rq was compatible with the Native Redis Backend configuration.

**Comment by obi-jerome** [comment]
@mohamed99elsokary
Before Django 4.0 we used django-redis to get a redis cache.
And django-rq could be simply configured then by setting "USE_REDIS_CACHE": "default".
Now we configure our cache with the new native Django Redis cache backend django.core.cache.backends.redis.RedisCache
But it is not declared as supported by django-rq, so we configured Redis manually for django-rq.

The native cache backend allows us to have one less dependency and to rely on the work of the Django Project that will probably give us a cache backend that is more reliable, integrated and performant, I guess.

It would be cool if django-rq was compatible with the Native Redis Backend configuration.

**Comment by CourteousCoder** [comment]
Hi. I was looking into this, and I think I found a way to integrate with Django's native Redis backend on Django 6 with `django.tasks` (related: #763)

I haven't tested this yet, but I'll let you know once I do. Happy to submit a PR too.

Here's the error message I got at the top of the stack trace.

```
connection_utils.py", line 41, in get_redis_connection
    return cache._client  # type: ignore[attr-defined]
           ^^^^^^^^^^^^^
AttributeError: 'RedisCache' object has no attribute '_client'
```

## TLDR

In `django_rq.connection_utils.get_redis_connection`, in addition to trying `cache._client`, we might also want to check one of these
```python
cache._cache.get_client(key=None, write=True)
# or 
cache._cache # handles serialization too
```

## Explanation

So, I looked at the source code and noticed that Django's native Redis cache backend doesn't point to the client directly with a `_client` attribute.
https://github.com/rq/django-rq/blob/39539ab680cf6a23073f4907b5e7332226494784/django_rq/connection_utils.py#L41

The the native Redis cache backend class is `RedisCache`, and it looks like just a thin wrapper around the client:

https://github.com/django/django/blob/5b3109329936e37432b1795f4b355ed60c4efd0a/django/core/cache/backends/redis.py#L173-L187

Internally, it has a property `_cache`, which returns a `RedisCacheClient`, which itself is a wrapper around the lower level [`redis.Redis` client](https://redis.readthedocs.io/en/stable/index.html) that handles serialization and connection pooling:

https://github.com/django/django/blob/5b3109329936e37432b1795f4b355ed60c4efd0a/django/core/cache/backends/redis.py#L32-L48

There's a `RedisCacheClient.get_client(self,key=None, write=False)` to get an instance of `redis.Redis` with it's connection pool. We might have to integrate serialization separately if we use this:
https://github.com/django/django/blob/5b3109329936e37432b1795f4b355ed60c4efd0a/django/core/cache/backends/redis.py#L93-L98

But I'm pretty sure `get_client` is what we need.

## Stack Trace

```bash
./manage.py check --fail-level=DEBUG 

Traceback (most recent call last):
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django_rq/connection_utils.py", line 41, in get_redis_connection
    return cache._client  # type: ignore[attr-defined]
           ^^^^^^^^^^^^^
AttributeError: 'RedisCache' object has no attribute '_client'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/chloe/myproject/./manage.py", line 23, in <module>
    main()
    ~~~~^^
  File "/home/chloe/myproject/./manage.py", line 19, in main
    execute_from_command_line(sys.argv)
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django/core/management/__init__.py", line 443, in execute_from_command_line
    utility.execute()
    ~~~~~~~~~~~~~~~^^
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django/core/management/__init__.py", line 437, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django/core/management/base.py", line 420, in run_from_argv
    self.execute(*args, **cmd_options)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django/core/management/base.py", line 464, in execute
    output = self.handle(*args, **options)
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django/core/management/commands/check.py", line 81, in handle
    self.check(
    ~~~~~~~~~~^
        app_configs=app_configs,
        ^^^^^^^^^^^^^^^^^^^^^^^^
    ...<4 lines>...
        databases=options["databases"],
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django/core/management/base.py", line 496, in check
    all_issues = checks.run_checks(
        app_configs=app_configs,
    ...<2 lines>...
        databases=databases,
    )
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django/core/checks/registry.py", line 95, in run_checks
    errors.extend(new_errors)
    ~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django/tasks/checks.py", line 11, in check_tasks
    yield from backend.check()
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django_tasks_rq/backend.py", line 281, in check
    django_rq.get_queue(queue_name)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django_rq/queues.py", line 146, in get_queue
    connection = get_connection(name)
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django_rq/connection_utils.py", line 85, in get_connection
    return get_redis_connection(QUEUES[name], use_strict_redis)
  File "/home/chloe/myproject/.venv/lib/python3.14/site-packages/django_rq/connection_utils.py", line 44, in get_redis_connection
    return cache.get_master_client()  # type: ignore[attr-defined]
           ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'RedisCache' object has no attribute 'get_master_client'
```

### `manage.py`

And for the sake of completeness, here's my `manage.py` file. Pretty standard.

```python3
#!/usr/bin/env python
"""Django's command-line utility for administrative tasks."""

import os
import sys


def main():
    """Run administrative tasks."""
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "myproject.core.settings.development")
    try:
        from django.core.management import execute_from_command_line
    except ImportError as exc:
        raise ImportError(
            "Couldn't import Django. Are you sure it's installed and "
            "available on your PYTHONPATH environment variable? Did you "
            "forget to activate a virtual environment?"
        ) from exc
    execute_from_command_line(sys.argv)


if __name__ == "__main__":
    main()
```

**Comment by CourteousCoder** [comment]
So this is the workaround workaround I came up with. I hope this helps someone!

```python3
# settings

CACHES = {
    "default": {
        "BACKEND": "myproject.cache.backends.RQCache",
        "LOCATION": "redis://localhost:6379/4",
    },
}

RQ_QUEUES = {
    "high": {
        "USE_REDIS_CACHE": "default",
    },
    "low": {
        "USE_REDIS_CACHE": "default",
    },
}

TASKS = {
    "default": {
        "BACKEND": "django_tasks_rq.RQBackend",
        "QUEUES": ["high", "low"],
    },
}

```

```python3
# myproject.cache.backends

from django.core.cache.backends.redis import RedisCache

class RQCache(RedisCache):
    """Customized Django RedisCache backend for Django RQ integration.
    :ref: https://github.com/rq/django-rq/issues/571#issuecomment-4568898583
    """

    @property
    def _client(self):
        """Get the master redis.Redis client
        When configuring an RQ_QUEUE to use USE_REDIS_CACHE, it looks for the redis.Redis client at self._client
        """
        return self._cache.get_client(write=True)
```
